modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -220,217 +158,220 @@ public void onContinuousProcessorStarted(GridKernalContext ctx) throws IgniteChe
-        }
+        registeredServices.clear();
 
-        ServiceConfiguration[] cfgs = ctx.config().getServiceConfiguration();
+        // If user requests sent to network but not received back to handle in exchange manager.
+        depFuts.values().forEach(fut -> fut.onDone(stopError));
+        depFuts.clear();
 
-        if (cfgs != null)
-            deployAll(Arrays.asList(cfgs), ctx.cluster().get().forServers().predicate()).get();
+        undepFuts.values().forEach(fut -> fut.onDone(stopError));
+        undepFuts.clear();
 
         if (log.isDebugEnabled())
-            log.debug("Started service processor.");
+            log.debug("Stopped service processor.");
     }
 
     /**
-     *
+     * Cancels deployed services.
      */
-    public void updateUtilityCache() {
-        serviceCache = ctx.cache().utilityCache();
-    }
+    private void cancelDeployedServices() {
+        assert opsLock.isWriteLockedByCurrentThread();
 
-    /**
-     * @return Service cache.
-     */
-    private IgniteInternalCache<Object, Object> serviceCache() {
-        if (serviceCache == null)
-            U.awaitQuiet(startLatch);
+        deployedServices.clear();
+
+        locServices.values().stream().flatMap(Collection::stream).forEach(srvcCtx -> {
+            cancel(srvcCtx);
+
+            if (ctx.isStopping()) {
+                try {
+                    if (log.isInfoEnabled()) {
+                        log.info("Shutting down distributed service [name=" + srvcCtx.name() + ", execId8=" +
+                            U.id8(srvcCtx.executionId()) + ']');
+                    }
+
+                    srvcCtx.executor().awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+                }
+                catch (InterruptedException ignore) {
+                    Thread.currentThread().interrupt();
+
+                    U.error(log, "Got interrupted while waiting for service to shutdown (will continue " +
+                        "stopping node): " + srvcCtx.name());
+                }
+            }
+        });
 
-        return serviceCache;
+        locServices.clear();
     }
 
     /** {@inheritDoc} */
-    @Override public void onKernalStop(boolean cancel) {
-        if (ctx.isDaemon())
+    @Override public void collectGridNodeData(DiscoveryDataBag dataBag) {
+        if (dataBag.commonDataCollectedFor(SERVICE_PROC.ordinal()))
             return;
 
-        GridSpinBusyLock busyLock = this.busyLock;
+        ServicesCommonDiscoveryData clusterData = new ServicesCommonDiscoveryData(
+            new ArrayList<>(registeredServices.values())
+        );
 
-        // Will not release it.
-        if (busyLock != null) {
-            busyLock.block();
+        dataBag.addGridCommonData(SERVICE_PROC.ordinal(), clusterData);
+    }
 
-            this.busyLock = null;
-        }
+    /** {@inheritDoc} */
+    @Override public void onGridDataReceived(DiscoveryDataBag.GridDiscoveryData data) {
```

Comment:
What if there is a conflict between local and received services? Will an error be logged?

Reply:
Any conflicts can't be here.
Joining node sends static configs in the cluster and receives verified and full data from the cluster, which will contain these static configs too if they pass validation.